### PR TITLE
CORS-3524, CORS-3523: aws: terraform: add spot instance support for masters

### PR DESCRIPTION
**Pull request by r4f4**
Spot instances can result on savings for short-lived clusters. If the control plane machine manifests have been edited with spot instance information, enable the use of spot instances in the terraform config. Notice that max price information is ignored, since it's not advised for it to be set.

### CODE REVIEW

**Comment by r4f4** [comment]
Depends on https://github.com/openshift/release/pull/51664 for testing spot instances deployments.

/hold

**Comment by r4f4** [comment]
/retitle CORS-3523: aws: terraform: add spot instance support for masters

**Comment by r4f4** [comment]
/retitle CORS-3524, CORS-3523: aws: terraform: add spot instance support for masters

File: cmd/openshift-install/features.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+var hiddenFeatures = [...]string{
+	"terraform-spot-masters",
+}
```
Comment (patrickdillon) [inline cmd/openshift-install/features.go:11]: I suppose grep works fine here, and perhaps we don't need to over engineer but would it be more convenient to structure this output, say in json or something?

Comment (r4f4) [reply cmd/openshift-install/features.go:11]: @patrickdillon what about a `--json` argument for that?

Comment (r4f4) [reply cmd/openshift-install/features.go]: Done

**Comment by r4f4** [comment]
/hold cancel
The way we're implementing this, https://github.com/openshift/release/pull/51664 and this PR can merge in any order.

File: cmd/openshift-install/features.go (hunk 1 above)
Comment (2uasimojo) [reply cmd/openshift-install/features.go]: Cool; though the current use case still requires:
- parsing the output with something (be it `grep` or `jq` or `[[ "$(...)" == *'"terraform-spot-masters"'* ]]`)
- accounting for what happens when invoking this on an older installer binary (stdout is empty, RC is nonzero)

...so YAGNI for now :shrug:

The only thing I can think of that would make it easier to consume would be something like `openshift-install is-hidden-feature-supported terraform-spot-masters` which exits zero if yes, nonzero if no. That way it can reliably be consumed as:

```bash
if ! openshift-install is-hidden-feature-supported terraform-spot-masters >/dev/null 2>&1; then
    ...
fi
```

...because I'll get a nonzero RC on older installers as well as newer ones that have this code but disclaim the support.

Comment (2uasimojo) [reply cmd/openshift-install/features.go]: to be clear: I'm not asking for :arrow_up:, just thinking aloud :)

**Comment by r4f4** [comment]
/override ci/prow/okd-images
Known issue

**Comment by 2uasimojo** [comment]
/lgtm cancel

Discussion still open around the hidden subcommand.

**Comment by patrickdillon** [comment]
> /lgtm cancel
> 
> Discussion still open around the hidden subcommand.

Yes, I am hesitant to add a new command to the installer for such a specific use. And I'm not certain that the problem requires a new command. I think it would help to have more context for the motivation. Reading through the slack thread I identified the motivation as:

> My goal was to put in a check so that, if you (accidentally or out of ignorance) tried to do this (tf + spot masters) on a .z without your PR, it would puke.

I am not sure if this is intended to be used in CI or for QE running manual tests or both. I would imagine in both of these cases we're pretty much only testing the latest z streams anyway, so my impression is that we're solving a transient problem with code that will live on much longer.

Can we layout in more detail the problem we're tackling with the command?

**Comment by 2uasimojo** [comment]
> Yes, I am hesitant to add a new command to the installer for such a specific use.

Ack, and likewise.

> And I'm not certain that the problem requires a new command.

I would love for us to come up with some other solution.

> Can we layout in more detail the problem we're tackling with the command?

If I pass an install-config with controlPlane spotMarketOptions into an installer with either CAPI featuregate flaggage or the terraform changes from this PR, it is honored.
If I pass an install-config with controlPlane spotMarketOptions into any other installer currently in existence, it is silently ignored.
Unchecked, this can easily result in a scenario where I *think* I'm saving money by enabling spot masters, but I'm actually not.

> I am not sure if this is intended to be used in CI or for QE running manual tests or both.

Yes, the primary use case is for CI jobs, where we have different personas responsible for the install-config vs the job config. It is via env vars to the latter that spot instance config (for masters and/or workers) is injected by tooling into the former. It is for this use case that we need protection, again because if I provide that env var but don't actually get spot masters:
- I haven't saved $$ as I thought I should.
- If something goes wrong, things could get confusing as I triage on the assumption that I'm using spot masters.

I imagine QE would find this similarly useful if they have scripting around generating install-configs. Not so much if they're doing everything by hand.

>  I would imagine in both of these cases we're pretty much only testing the latest z streams anyway, so my impression is that we're solving a transient problem with code that will live on much longer.

This is a sane point. If we intend to backport this change to all living y-streams, *and* CI/QE exclusively tests *only* the latest (nightlies and) z-streams thereon, I agree the window of usefulness is smallish. Are those assumptions valid? In particular, do we really not run CI on old-but-still-supported z-streams?

**Comment by r4f4** [comment]
> > /lgtm cancel
> > 
> > Discussion still open around the hidden subcommand.
> 
> Yes, I am hesitant to add a new command to the installer for such a specific use. And I'm not certain that the problem requires a new command. I think it would help to have more context for the motivation. Reading through the slack thread I identified the motivation as:
> 
> > My goal was to put in a check so that, if you (accidentally or out of ignorance) tried to do this (tf + spot masters) on a .z without your PR, it would puke.
> 
> I am not sure if this is intended to be used in CI or for QE running manual tests or both. I would imagine in both of these cases we're pretty much only testing the latest z streams anyway, so my impression is that we're solving a transient problem with code that will live on much longer.
> 
> Can we layout in more detail the problem we're tackling with the command?
> 

That's not totally true because we have to modify the IPI steps which are shared among all ocp versions. So we need a why to identify which versions of the installer support the spot instances feature (the other option is that older installers will silently ignore the spot feature, even though the master manifests were changed. That's what we were trying to avoid).

**Comment by patrickdillon** [comment]
@r4f4 @2uasimojo what about solving these problems with tests? In ci run a post-install test that checks the envvar and if set ensures spot instances are running. Share this tactic with qe

**Comment by patrickdillon** [comment]
> @r4f4 @2uasimojo what about solving these problems with tests? In ci run a post-install test that checks the envvar and if set ensures spot instances are running. Share this tactic with qe

sorry was in a rush to share this idea but I realize I did it in an entirely unconvincing and simple way. i'm not sure how large the surface area in ci is for failing this test, but if it is large, the test would need to be optional to be informing (this would give us an idea of how to expand coverage), if it is massive perhaps this idea would not work. this is perhaps less of a silent fail than the proposed solution is?

**Comment by r4f4** [comment]
> > @r4f4 @2uasimojo what about solving these problems with tests? In ci run a post-install test that checks the envvar and if set ensures spot instances are running. Share this tactic with qe
> 
> sorry was in a rush to share this idea but I realize I did it in an entirely unconvincing and simple way. i'm not sure how large the surface area in ci is for failing this test, but if it is large, the test would need to be optional (not sure if this is possible) to be informative (this would give us an idea of how to expand coverage), if it is massive perhaps this idea would not work. this is perhaps less of a silent fail than the proposed solution is? 

Remember that backports in the installer take way longer than we wish they would. So I think failing tests is out of question because we'd fail too many. Besides, a mechanism for detecting available features would be useful for other ci-only functionality like public-only subnets and BYOIPv4 (which is active for certain regions but we are silently ignoring until we add support in the capi path).

It doesn't need to be a hidden command but I think it'll be very useful to have something.

**Comment by 2uasimojo** [comment]
> solving these problems with tests? In ci run a post-install test that checks the envvar and if set ensures spot instances are running.

Interesting idea. Digging in...

IIUC this would look like:
Include a post-install step in the IPI flow where we're adding support for spot masters.
The post-install step would, if `SPOT_MASTERS` is set, inspect the masters on the created cluster and make sure they're spots, failing the setup if they're not.
To be useful, this step would need to **always** be part of the ipi setup flow.
It would thus affect, what, hundreds? of CI jobs.
...which, if misconfigured, would be creating entire clusters -- with expensive on-demand instances -- and quickly destroying them (without running the meat of whatever broader test this is part of) because they're misconfigured.
And these test would take ~40m to fail, vs "immediately".
I suppose we would catch this misconfig for a given job at pj-rehearse time... assuming we ran a pj-rehearse for every job we try to switch this on for, which doesn't sound like it would be reasonable for the scale we're talking about.
Also, anecdotally, the rate at which job failures are ignored in OSCI... I'm not even sure we would catch or react to this with any efficiency.

In summary: IMHO making the failure fast and cheap is moar better than making it slow and expensive.

**Comment by patrickdillon** [comment]
> > solving these problems with tests? In ci run a post-install test that checks the envvar and if set ensures spot instances are running.
> 
> Interesting idea. Digging in...
> 
> IIUC this would look like: Include a post-install step in the IPI flow where we're adding support for spot masters. 

 write/run a test--not add a step. I haven't written such a test before so I don't know how it's done offhand (would need to nail this down to be serious).  the test would fail if the job config sets the envvar for spot instances but the resulting cluster isn't running (on) them. Ideally would be able to use something like ci search or sippy to identify failing jobs.

But I agree that it is a much slower feedback loop.  This discussion helped me understand how we're going to use this to roll out the spot instances.  

/approve
@2uasimojo feel free to re-lgtm we are probably going to need to backport this pretty far I assume

**Review by patrickdillon — APPROVED** [review]

**Comment by 2uasimojo** [comment]
Awesome, let’s do it

/lgtm

**Comment by 2uasimojo** [comment]
/cherry-pick release-4.15 release-4.14 release-4.13 release-4.12

**Comment by r4f4** [comment]
/retitle OCPBUGS-34975: aws: terraform: add spot instance support for masters
